UI Dropdown

alex40724 opened the pull request.

gpkvt committed:
Converted ilServer README.txt to README.md

gpkvt committed:
Updated path to ilServer README

gpkvt committed:
Update INSTALL.md

Commit:
Filesystem interface created.

Commit:
Visibility interface added to make the usage of the filesystem rights more readable.

Commit:
Change log and readme added.

Commit:
Merge branch 'trunk' into feature/5-3/FilesystemService

Commit:
Migrated interfaces down to PHP 5.6

Commit:
changed heading of the change log because the log belongs to the service and not to the project

Commit:
First version of the FileUpload interface.

Commit:
Doc comments updated and type checking code duplicates eliminated.

Commit:
EntryLockingStringMap and ImmutableMapWrapper added.

Commit:
Change log added.

Commit:
Size renamed to upload size which should be more expressive.

Commit:
Preprocessors are new using the psr-7 stream interface. This change provides the processor developers with a more object oriented way to deal with the file stream and decouples the processors from the php resource it self.

Commit:
Changed the moveFilesTo method to enable the developers to actually chose the filesystem they want to use for the files. The default is the ILIAS storage filesystem.

Commit:
Readme added which outlines the current interface of the file upload service.

Commit:
Rename UploadStatus to ProcessingStatus because the status belongs to the processors and not to the upload it self.

Commit:
UploadStatus renamed but SourceTree staged only half of the operation.

Commit:
PreProcessorManager added which allows the usage of the preprocessors without the file upload service. This is useful for unzip and chron jobs which are dealing with the files but not with the upload process it self.

Commit:
Interface split and updated

Commit:
Readme updated

Commit:
Change log updated

Commit:
Extend readme of the UI-Service with some information about the javascript abstraction.

Commit:
IllegalStateException added.

Commit:
replaced IllegalArgumentException with the native InvalidArgumentException

Commit:
filesystem interfaces updated

Commit:
The rest of the previews commit because something went wrong while staging the previews changes.

Commit:
Filesystem factories added which enables the ilInitialisation to generate new filesystem instances.

Commit:
Filesystems interface added which is used within the ILIAS container.

Commit:
Cross filesystem operation description added.

Commit:
Change log updated.

chfsx committed:
updated PHPUnit to 5.7

chfsx committed:
replaced TestCase::getMock with ::createMock

chfsx committed:
Merge branch 'trunk' into proposal/update-phpunit-5_7

* trunk:
  refactored identifiers for better understanding (since i needed to think about what it is ^^) and fixed the issue of wrong context for output causing the stripping of input html that is actually required but only accepted as non input html (for presentration only)
  Test: Hopefully fixed #19874

chfsx committed:
added class-map for /Services and /Modules

chfsx committed:
removed all requires in DCL

chfsx committed:
removed all requires in DCL

chfsx committed:
Optimized Classmap

chfsx committed:
removed requires ind BIBL

xus committed:
Remove ###deprecated lang vars

Using this
tool:https://github.com/ILIAS-eLearning/DeveloperTools/blob/master/mark_
legacy_lang_vars/remove_marked_deprecates.py

chfsx committed:
optimize autoloader activated

chfsx committed:
Revert "removed requires ind BIBL"

This reverts commit bb8e3c560b3e0a9cbd9e061ca6d2dc2e069b17d3.

chfsx committed:
Revert "removed all requires in DCL"

This reverts commit 446cc304957254f0e202ccd33c761d7fe12ca477.

chfsx committed:
PHP dependencies

nhaagen committed:
data type: color

nhaagen committed:
data type: color

Commit:
ks drop down implementation proposal

Commit:
ks drop down implementation proposal

Amstutz commented on src/UI/Component/DropDown/DropDown.php:
```diff
@@ -0,0 +1,31 @@
+
+/* Copyright (c) 2017 Alexander Killing <killing@leifos.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\DropDown;
```
Dropdown is a single word, so camel case is not needed here. Replace DropDown with Dropdown.

alex40724 replied:
ok

Amstutz commented on src/UI/Component/DropDown/DropDown.php:
```diff
@@ -0,0 +1,31 @@
+use ILIAS\UI\Component\JavaScriptBindable;
+
+/**
+ * This describes commonalities between all types of drop downs
```
Handle dropdown as one word, as bootstrap does: http://getbootstrap.com/javascript/#dropdowns

alex40724 replied:
Got it :-)

Amstutz commented on src/UI/Component/DropDown/DropDownItem.php:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Drop down items
+ */
+interface DropDownItem extends Component, JavaScriptBindable, Clickable, Hoverable {
```
I am not sure, if this is placed correctly here. I dislike the idea of a DropdownItem. The structure of the item could be reusable. Just passing an array could be easier (similar the the descriptive listing). However, you will surely dislike the idea of not being able to perform a good check on array. This is valid. The Current version of this also does not solve the issue of not having dividers. What about the following:
```
$items = array(
		$f->link("ILIAS", "https://www.ilias.de"),
                $f->divider(),
		$f->link("GitHub", "https://www.github.com")
	);
return $renderer->render($f->dropdown()->standard($items)->withLabel("Actions"));
```
This would imply, that we need links and dividers in the UI Framework... Not sure if this would be the way forward. What do you think @alex40724 and @klees?

alex40724 replied:
Or we have DividerItem and LinkItem being derived from DropdownItem and a DropDown constructor with (DropDownItems[] $items) being passed? Or a Link and Divider on a higher level implementing a DropdownItem interface? I would prefer anything that I could type hint in the signature in the future.

In your case Link and Divider would be independent components? So an iteration on the array and type check of each item would be necessary?

I think we definitely will need the JS/ID magic on the items.

Amstutz replied:
> I would prefer anything that I could type hint in the signature in the future.

+1

> In your case Link and Divider would be independent components? So an iteration on the array and type check of each item would be necessary?

Not really. The link and the divider would render themselves in the context dropdown. See https://github.com/ILIAS-eLearning/ILIAS/pull/532

alex40724 replied:
So no type checking? Just calling render() on all items?

If we define contexts on this level of granularity, we might end up with quite a complex context stack. And either I got #532 wrong (most probably) or this won't really work imo.

Amstutz replied:
I got your question wrong sorry. On the component you would to a type check, but not on the renderer.

Concerning the context stack. That might be an issue. However, why would it not be possible?

alex40724 replied:
OK, I suggest to continue the context discussion in #532. But how to proceed with link and divider? This Dropdown proposal is a result of the AppointmentItem of the Calendar Revision. I think we might need JS on the action items quickly and working with the original label/action array only, won't give us this (in my understanding). What if one of these drop down items should open a Modal?

On the other hand we do not have a Link or Divider ready. Should we try to provide this? Would you start PRs on them? I currently do not feel to comfortable with the hierarchy... We get lots of dependencies here. I am not sure how fast we find agreement on these things.

Amstutz replied:
Had some thoughts about this. 

**Concerning the dividers:**
- I would suggest now to not introduce them as own item. There is not much to be gained there. 
- Much more I would introduce a new method on the dropdown itself enabling to add individual sections, e.g. addSection($items). The code to generate with two section would look as follows:
```
$myDropdownWithTwoSections = $f->dropdown()->standard(
     [$entry1section1,$entry2section1]
)->addSection([$entry1section2,$entry2section2]);
```
**Concerning the items:**
- I would still suggest to add links and not dropdown items. To me, those entries are exactly that. I think you would not even need a special renderer for that (since we do not need to change the DOM). Bootstrap/ILIAS can do everything needed through CSS. This also separates dropdowns from Popovers, dropdowns only hold links, Popovers may also contain other elements. One could argue with this separation but it is to some degree an adaption of the bootstrap terminology.
- Concerning your question about an item opening a specific modal, this would look as follows with links:

```
$modal = $factory->modal()->roundtrip('My Modal', $content));
$link = $factory->link()->standard('Open Modal 1', '#')
    ->withOnClick($modal->getShowSignal());
$f->dropdown()->standard( [$link]);
```



- Concerning the question, who will Propose/Implement the links, I could do that.

@alex40724, @klees Major objections to this proposal? If yes we can stick with the dropdown items. In this case I would still suggest to add the addSection Method for the individual sections with dividers.

alex40724 replied:
Oha, Dividers. Just used your idea in #550, so in the meantime I would support it and also have some kind of implementation ready. I think handling of sections would be more complicated (for the appointment lists you refused some "handling of grouping there", if I remember correctly). From the UI point of view we deal with a sequence of links and dividers, so I have a preference to implement your initial proposal now. :-)

I can implement Links and Dividers, if we agree on where to place it in the hierarchy. Both top level?

alex40724 replied:
Top > Link
Top > Divider

klees replied:
Hi @alex40724, @Amstutz,

@nhaagen is currently working on the breadcrumb, and, surprise, also found the concept of a link. I.e. the breadcrumbs are nothing more than a specialized list of links, very similar to the dropdown in that regard. So maybe a "link list" could be a node in the hierarchy, with the children "dropdown" and "breadcrumb".
@nhaagen also, imo correctly, observed, that a "link" (i.e. a tuple of URL and title) seems to be a data type in its own right, independent from the UI framework, as it should also be usable in other contexts. That means we might want to introduce it in src/Data instead of the UI-framework. This would also solve the context problem. What do you think? We are able to propose an implementation for Data now, if we agree to that solution.
Regards!

Amstutz replied:
@klees I am not sure if "instead" is correct here. If we only have links in src/Data and therefore can have no coupled events (e.g. Click). How would you solve this?

alex40724 replied:
I definitely need an UI Link component due to the Javascript binding/signaling features as written above. And in contrast to the Divider I think a renderer for a "stand alone" link would also be feasible. We might limit the use of where links can appear drastically by rules for now.

alex40724 replied:
Link PR: https://github.com/ILIAS-eLearning/ILIAS/pull/552

Amstutz commented on src/UI/Component/DropDown/Factory.php:
```diff
@@ -0,0 +1,44 @@
+	 *       Drop down items represent single entries in a drop down.
+	 *   composition: >
+	 *       Drop down items cannot be rendered on their own, instead they are passed
+	 *       to a \ILIAS\UI\Component\DropDown\DropDown instance.
```
This indicates to me, that they should not get their own entry.

alex40724 replied:
Isn't this more or less the same as with Counters?

Amstutz replied:
To some degree yes. However, they have own renderer and template. But if I think about it, yes, it is in it's essence almost the same.

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *      a different view.
+	 *   composition: >
+	 *      Drop Down is a clickable, graphically obtrusive control element. It can
+	 *      bear text.
```
and horizontal dividers.

alex40724 replied:
I definitely agree that this should be possible.

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *      bear text.
+	 *   effect: >
+	 *      On-click, a list of actions is revealed. Clicking an item will trigger the action indicated.
+	 *      Clicking outside of an opened drop down will close the list of items.
```
Is there any async action possible?

alex40724 replied:
For loading the items. Not. But this is not needed for the first use case (appointment lists) imo.

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *      links: >
+	 *          Links are used to trigger Interactions that do not change the systems
+	 *          status. They are usually contained inside a Navigational Collection.
+	 *
```
Maybe also add popovers here: Dropdowns always display a list of things while Popovers are much more flexible in the things they can display.

alex40724 replied:
And SplitButtons, right? I was wondering in the beginning if a Dropdown is a kind of Button. Still I am not sure about this. But somehow I think these are different things.

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *           of four or fewer letters MUST be capitalized.
+	 *      3: >
+	 *           For standard events such as saving or canceling the existing standard
+	 *           terms MUST be used if possible: Save, Cancel, Delete, Cut, Copy.
```
Good!

alex40724 replied:
...same.

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *           The caption of a Drop Down SHOULD contain no more than two words.
+	 *      2: >
+	 *           Every word except articles, coordinating conjunctions and prepositions
+	 *           of four or fewer letters MUST be capitalized.
```
Good!

alex40724 replied:
Copy/pasted from the Button component...

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *           wide.
+	 *   wording:
+	 *      1: >
+	 *           The caption of a Drop Down SHOULD contain no more than two words.
```
Good!

alex40724 replied:
But the "must not be used on their own" part is also related to the "DropdownItems have their own entry" issue, or not?

Amstutz commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,63 @@ public function panel();
+	 *           element as a Drop Down.
+	 *      2: >
+	 *           Drop Down Items are being implemented as "ul" list with a set of "li" elements and
+	 *           nested "a" elements for the actions.
```
I have some additions for accessability:

- Triggers of Dropdowns MUST indicate their effect by the aria-haspopup attribute set to true.
- Triggers of Dropdowns MUST indicate the current state of the Dropdown by the aria-expanded label.
- Dropdowns MUST be accessible by keyboard by focusing the trigger element and clicking the return key. 
- Entries in a Dropdown MUST be accessible by the tab-key if opened.
- The focus MAY leave the Dropdown if tab is pressed while focusing the last element. This differs from the behaviour in Popovers and Modals.

alex40724 replied:
Thanks, I will check all this.

Amstutz commented on src/UI/Component/DropDown/Factory.php:
```diff
@@ -0,0 +1,44 @@
+	 * @param \ILIAS\UI\Component\DropDown/DropDownItem[] array of action items
+	 * @return \ILIAS\UI\Component\DropDown\Standard
+	 */
+	public function standard($items);
```
Dropdowns without label are possible, correct? They just contain the caret?

alex40724 replied:
Yes, this is the idea.

Commit:
Removed unused import.

nhaagen committed:
added isDark-function to determine luminance

nhaagen committed:
removed obsolete vars

Commit:
Moved FileSize to Data namespace and renamed class to DataSize

Amstutz committed:
UI: Removed unused and wrongly placed table class.

Commit:
Cloud: avoid locator bug with cloud objects

Amstutz committed:
System Styles: Fix of 20661.

Commit:
Metadata documentation updated and MetadataType constants added to eliminate "magic values".

Commit:
hasDir method to DirectoryAccess added.

nhaagen committed:
doc-notes

Commit:
Filesystem access interfaces split.

Commit:
Some of the interfaces had a wrong documentation regarding the type of the Interface.

Commit:
Own FileStream interface added which is still compatible with the http service streams.

Commit:
README adjusted.

Commit:
Change log updated.

Commit:
fixed missing functions and function calls

Commit:
fixed missing functions and function calls

Commit:
fixed html

Commit:
fixed html

Commit:
docu fixes

Commit:
docu fixes

Amstutz reviewed (changes requested):
First of all, thank you a lot for this PR! 

Four main things:
- Items: Not sure if the taxonomy is correct. See according comment for details.
- Dividers: We need those, to make the dropdowns easier to read.
- Async: We probably need Dropdowns with async loaded entries pretty much from the start. Fortunately, all logic we need for this exists already (see popovers PR). I can help with this, if needed.
- Tests: Please add Unit Tests.

alex40724 commented:
@Amstutz Thanks for the review.

- Are missing unit tests a reason to decline KS PR at this stage?
- I also think it should be possible that async loading of items could be implemented later, since this need is rather exceptional. The actions in the Appointment Lists didn't come with this, too.
- I agree that we need to make a decision on the divider/items issues.

Commit:
0020403: Loss of Data

smeyer-ilias committed:
0019460: Failed test: Automatisches Auffüllen in Kurs durch selbständigen Kursaustritt auslösen

Commit:
0020576: umlauts can't be compared (by importing .xls files)

Commit:
central pass determination considering previous solutions within test player, pass index to be used is passed to assQuestionGUI rendering for sure, rendering considers intermediates only when previous pass was not determined (determination results gets flagged in question) - minimal but qualified patch also cleaning up some issues with subnav and javaapplet

Commit:
small fixes, missing template

Commit:
small fixes, missing template

Commit:
0020626: Invisible fields visible in notification mail

Commit:
0020609: Excel import fails

Commit:
0020609: Excel import fails (2)

Commit:
0020633: Datenfilter-Problem in Datensammlung

Commit:
fixed referencing the new template

Commit:
fixed use of wrong lang variable

Commit:
0020700: Visibility of main tables

Amstutz commented:
Hi Alex

> Are missing unit tests a reason to decline KS PR at this stage?

- For the JF, no absolutely not. This already goes far beyond what is asked to be discussed in the JF. See: https://github.com/ILIAS-eLearning/ILIAS/tree/trunk/src/UI#how-to-implement-a-component step 6. I am only paranoid, that those tests could be forgotten :-). Maybe we need to define a little closer what "Reject" at various stages means. 
- For a merge, yes.

> I also think it should be possible that async loading of items could be implemented later, since this need is rather exceptional. The actions in the Appointment Lists didn't come with this, too.

Agreed for the async part, we can do that later. It is already great, that you propose this.

> I agree that we need to make a decision on the divider/items issues.

See comment above.

Commit:
fixed mangtis 0020671: Question preview is corrupted +++ cleaned up domain and usage of rendering purpose constants (former outputmode) for the common qGUI outputs (preview/testoutput/solutionoutput)

smeyer-ilias committed:
0020702: \ilSoapLearningProgressAdministration::getProgressInfo: Undefined constant LP_MODE_UNDEFINED

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

Commit:
fixed awareness tool for members of my groups/courses, 20479: My Groups and Courses: online and offline doesn't show any Users

Commit:
fixed wrong object access after moving method

Commit:
fixed bug #18616: Auto-linked Glossary terms should not be linked when inserting source code in a Learning Module Page

Commit:
fixed bug #20679: Snippets in copied Media Pools fail

Commit:
fixed bug #20628: error occurs with colors in data table (JavaScript-Modus)

Commit:
fixed bug #20478: Import von ILIAS Lernmodulen schlägt fehlt

Commit:
fixed bug 20706: Failed Import Course with learning objectives

Commit:
added logging to membership news cron job

nhaagen committed:
factory for different ways of constructing a color

nhaagen committed:
color stores r,g,b instead of hex, moved validation/nomalization to factory

nhaagen committed:
test single color getters

nhaagen committed:
updated doku.

nhaagen committed:
cache color-factory

smeyer-ilias committed:
Fixed duplicate lang entry

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

Commit:
fixed selected file reusing with previous solution adoption and file upload questions // refactored a lot of small common stuff (e.g. lookups)

Commit:
Merge branch 'fixSubNavPreviousSolutionsPatch' into fixSubNavPreviousSolutionsPatch_53x

Commit:
added debug messages; optimize array usage

matthiaskunkel committed:
Updated URL to new MathJax server after cdn.mathjax.org server has been shut down.

chfsx committed:
some code-reformat

matthiaskunkel committed:
Replaced outdated mathjax server info and added new one

chfsx committed:
Updated readme

matthiaskunkel committed:
Fixed Mantis #20644 - typo in auth#:#auth_radius_sync_rad_info#

Commit:
0020722: Bad representation of file information

matthiaskunkel committed:
Fixed Mantis #20516 and #20515, typos in survey

smeyer-ilias committed:
fixed typo

Commit:
Changed default Url for MathJax (fix 20409 and 20638)

chfsx committed:
Merge branch 'trunk' into proposal/classmap

* trunk: (155 commits)
  Changed default Url for MathJax (fix 20409 and 20638)
  fixed typo
  0020722: Bad representation of file information
  Fixed Mantis #20516 and #20515, typos in survey
  Fixed Mantis #20644 - typo in auth#:#auth_radius_sync_rad_info#
  Replaced outdated mathjax server info and added new one
  Updated URL to new MathJax server after cdn.mathjax.org server has been shut down.
  added debug messages; optimize array usage
  Fixed duplicate lang entry
  added logging to membership news cron job
  fixed bug 20706: Failed Import Course with learning objectives
  fixed bug #20478: Import von ILIAS Lernmodulen schlägt fehlt
  fixed bug #20628: error occurs with colors in data table (JavaScript-Modus)
  fixed bug #20679: Snippets in copied Media Pools fail
  fixed bug #18616: Auto-linked Glossary terms should not be linked when inserting source code in a Learning Module Page
  fixed awareness tool for members of my groups/courses, 20479: My Groups and Courses: online and offline doesn't show any Users
  0020702: \ilSoapLearningProgressAdministration::getProgressInfo: Undefined constant LP_MODE_UNDEFINED
  fixed mangtis 0020671: Question preview is corrupted +++ cleaned up domain and usage of rendering purpose constants (former outputmode) for the common qGUI outputs (preview/testoutput/solutionoutput)
  0020700: Visibility of main tables
  0020633: Datenfilter-Problem in Datensammlung
  ...

chfsx committed:
updated classmap

Amstutz committed:
Merge pull request #523 from studer-raimann/feature/5-3/readme-ui-service-signals

Add section about javascript abstraction to readme of UI-Service

smeyer-ilias committed:
Fixed 0020507: Several issues while joining a course with password

alex40724 commented:
Top > Dropdown ok for now.

nhaagen committed:
renamed validateRGB/Hex to checkRGB/Hex

klees committed:
Merge pull request #542 from nhaagen/DataType_Color

Data type "color"

klees committed:
Data/Color: minor refactoring of tests.

klees committed:
Data: moved Color to Data-namespace.

mjansenDatabay committed:
Auth: Fixed #20685

mjansenDatabay committed:
Auth: Fixed #20685

Commit:
compatibility to 52x merge (due to pickable cherries)

smeyer-ilias committed:
Merge branch 'patch-1' of https://github.com/gpkvt/ILIAS into gpkvt-patch-1

Conflicts:
	docs/INSTALL.md

smeyer-ilias committed:
Merge branch 'gpkvt-patch-1' into trunk

Commit:
imagemap compatibility niveau

Commit:
Merge branch 'fixSubNavPreviousSolutionsPatch_53x' into trunk

Commit:
ks link proposal

Commit:
fixed remove selection request detection (selection removal broken)

nhaagen committed:
added example for color

nhaagen committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into DataType_Color

Commit:
fixed confusing interaction when modifying prefilled imagemap solutions

klees committed:
Merge pull request #553 from nhaagen/DataType_Color

Data, Color: added example for color

nhaagen committed:
donot use namespace doubly

nhaagen committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into DataType_Color

klees committed:
Merge pull request #555 from nhaagen/DataType_Color

Datatype Color  - fix double use of namespace in Readme

klees committed:
Data: changed examples in README a bit, hopefully this makes hhvm happy.

Commit:
shy buttons

Commit:
shy buttons

Commit:
added missing less and template file

Commit:
reduced link concept

chfsx committed:
changes after new components.json from Fabian Kruse

chfsx committed:
removed documentation of unmaintained directories

Commit:
ks divider

smeyer-ilias committed:
0020588: Can't use multiple Radius-Servers for round robin

mjansenDatabay committed:
Mail: Fixed #20746

mjansenDatabay committed:
Merge remote-tracking branch 'origin/trunk' into trunk

smeyer-ilias committed:
0020739: Deleting Entrytest or graduating Test leads to Error in course with learning objectives

smeyer-ilias committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into trunk

chfsx committed:
[FIX] 0020755: Upload a zip-file with unzip => Non-static method

smeyer-ilias committed:
0019460: Failed test: Automatisches Auffüllen in Kurs durch selbständigen Kursaustritt auslösen ii

klees committed:
adjusted dicto-rules for new dicto version.

Commit:
rule clarifications

smeyer-ilias committed:
0017561: Adopting content works, but progress bar gets stuck at 100%

Commit:
updated dropdown interface: removed dropdown items, used shy buttons and dividers, fixed naming

Commit:
updated dropdown interface: removed dropdown items, used shy buttons and dividers, fixed naming

alex40724 commented:
Major changes with the latest commit:
- Removed Dropdown items, use Shy Buttons and Dividers instead
- Fixed naming ("Dropdown" instead of "DropDown" or "Drop down"

Implementation needs fixes, only the interface is up-to-date.

Major open decision: Should it be
Dropdown > Standard
or
Button > Dropdown > Standard
?

smeyer-ilias committed:
0020758: Menüpunkt wird aktiv, obwohl nicht angeklickt

klees commented on src/UI/Implementation/Component/Dropdown/Factory.php:
```diff
@@ -0,0 +1,16 @@
+	 */
+	public function standard($items) {
+        return new Standard($items);
+    }
```
Indentation!

alex40724 commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,80 @@ public function panel();
+	 *      button: >
+	 *          Buttons are used, if single actions should be presented directly in the user interface.
+	 *      links: >
+	 *          Links are used to trigger Interactions that do not change the systems
```
... action ...

alex40724 commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,80 @@ public function panel();
+	 *          Links are used to trigger Interactions that do not change the systems
+	 *          status. They are usually contained inside a Navigational Collection.
+	 *      popovers: >
+	 *          Dropdowns only list a list of possible actions. Popovers can include more diverse
```
provide a list

alex40724 commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,80 @@ public function panel();
+	 * rules:
+	 *   usage:
+	 *      1: >
+	 *           Dropdowns MUST NOT be used on their own. They are only parts of more complex UI elements.
```
standalone

alex40724 commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,80 @@ public function panel();
+	 *           wide.
+	 *   wording:
+	 *      1: >
+	 *           The caption of a Dropdown SHOULD contain no more than two words.
```
label

alex40724 commented on src/UI/Factory.php:
```diff
@@ -336,4 +336,80 @@ public function panel();
+	 *           of four or fewer letters MUST be capitalized.
+	 *      3: >
+	 *           For standard events such as saving or canceling the existing standard
+	 *           terms MUST be used if possible: Save, Cancel, Delete, Cut, Copy.
```
get rid of Save and Cancel

chfsx committed:
Merge pull request #519 from studer-raimann/feature/5-3/FileUploadService

Feature/5 3/file upload service

alex40724 reviewed (commented):
JF Dropdown > Standard

matthiaskunkel commented:
Jour Fixe, June 19, 2017: Pull request accepted. Please merge to trunk.

chfsx committed:
Merge branch 'trunk' of https://github.com/ILIAS-eLearning/ILIAS into proposal/classmap

Amstutz committed:
Lng: Added missing variables.

smeyer-ilias committed:
Added fix for 0020760: NIC-ID not given

klees committed:
Data: replaced asserts by if/throw to make hhvm happy.

Amstutz committed:
System Styles: Improved labels

matthiaskunkel committed:
Merge pull request #539 from xus/trunk_remove_deprecateds_lang

Remove ###deprecated lang vars from trunk

chfsx committed:
moved composer README

chfsx committed:
Added info for classmap generation

chfsx committed:
updated classmap

chfsx committed:
Merge pull request #541 from studer-raimann/proposal/classmap

Proposal/classmap

chfsx committed:
Merge branch 'trunk' into proposal/update-phpunit-5_7

* trunk: (229 commits)
  updated classmap
  Added info for classmap generation
  moved composer README
  System Styles: Improved labels
  Data: replaced asserts by if/throw to make hhvm happy.
  Added fix for 0020760: NIC-ID not given
  Lng: Added missing variables.
  0020758: Menüpunkt wird aktiv, obwohl nicht angeklickt
  0019460: Failed test: Automatisches Auffüllen in Kurs durch selbständigen Kursaustritt auslösen ii
  0017561: Adopting content works, but progress bar gets stuck at 100%
  adjusted dicto-rules for new dicto version.
  [FIX] 0020755: Upload a zip-file with unzip => Non-static method
  0020739: Deleting Entrytest or graduating Test leads to Error in course with learning objectives
  Mail: Fixed #20746
  0020588: Can't use multiple Radius-Servers for round robin
  removed documentation of unmaintained directories
  changes after new components.json from Fabian Kruse
  Data: changed examples in README a bit, hopefully this makes hhvm happy.
  donot use namespace doubly
  fixed confusing interaction when modifying prefilled imagemap solutions
  ...

chfsx committed:
Merge pull request #533 from studer-raimann/proposal/update-phpunit-5_7

Proposal/update phpunit 5 7

chfsx committed:
removed PHP-version.dependecies since our CI isn't abe to handle them.

chfsx committed:
Merge branch 'trunk' into feature/5-3/FilesystemService

* trunk: (277 commits)
  removed all requires in DCL
  removed PHP-version.dependecies since our CI isn't abe to handle them.
  updated classmap
  Added info for classmap generation
  moved composer README
  System Styles: Improved labels
  Data: replaced asserts by if/throw to make hhvm happy.
  Added fix for 0020760: NIC-ID not given
  Lng: Added missing variables.
  0020758: Menüpunkt wird aktiv, obwohl nicht angeklickt
  0019460: Failed test: Automatisches Auffüllen in Kurs durch selbständigen Kursaustritt auslösen ii
  0017561: Adopting content works, but progress bar gets stuck at 100%
  adjusted dicto-rules for new dicto version.
  [FIX] 0020755: Upload a zip-file with unzip => Non-static method
  0020739: Deleting Entrytest or graduating Test leads to Error in course with learning objectives
  Mail: Fixed #20746
  0020588: Can't use multiple Radius-Servers for round robin
  removed documentation of unmaintained directories
  changes after new components.json from Fabian Kruse
  Data: changed examples in README a bit, hopefully this makes hhvm happy.
  ...

Commit:
added tests for link

chfsx committed:
Merge pull request #526 from studer-raimann/feature/5-3/FilesystemService

Feature/5 3/filesystem service

Commit:
0020769: using survey-reminder-temptes fails if subject is empty

Commit:
renamed standard to horizontal, clarified description, added tests, fixed less

Commit:
removed comment

Commit:
added tests

Commit:
0017842: Kann nicht das PluginMenü unter Administration aufrufen

Commit:
Merge pull request #556 from leifos-gmbh/5_3_cal_ks_shy_button

UI Shy Button

Commit:
fixed description and less variables

Commit:
merged UI Divider branch

Commit:
merged UI Link branch

Commit:
fixed Factory and data.json file

Commit:
fixed factories and data.json
